### ISSUE TITLE
[feature] Search for documents using document/object/dict fragments

### ISSUE THREAD

**Issue by munael** [msiemens/tinydb#300]
Basically, something like this:

```py
db.search({'key1': 'val1', 'key7': var7})

# Translates to
db.search((q.key1 == 'val1') & (q.key7 == var7))
```

**Comment by narfanar** [comment]
As a start:

```py
def search_d(self, fragment: dict):
    q = Query()
    expr = True
    for k, v in fragment.items():
        expr = expr & (getattr(q, k) == v)
    return self.search(expr)
```

**Comment by msiemens** [comment]
Hey @narfanar,

thanks for your feature request! It's a nice idea but I wonder if enough people would use it to make it worth including in the TinyDB core. I'm open to implementing this if other people think they would use it too.

**Comment by corydodt** [comment]
I know anecdotes are data, but since you asked:

I just discovered and implemented tinydb today in my app and this feature was one of the first ones I wished for. :)

**Comment by yuban10703** [comment]
```python
from tinydb import TinyDB, Query
from tinydb.storages import MemoryStorage


tmpDB = TinyDB(storage=MemoryStorage)
Q = Query()
tmpDB.insert({'level': 'sexy', 'tags': [{'name': '愛のあるセックス'}, {'name': '有爱的性'}, {'name': '裸ワイシャツ'}, {'name': '裸体衬衫'}]})

tmpDB.insert({'level': 'sexy', 'tags': [{'name': 'お腹ナデナデ'}, {'name': 'belly rub'}, {'name': '下脱ぎ'}, {'name': '裸下半身'}]})

tmpDB.insert({'level': 'sexy', 'tags': [{'name': 'オリジナル'}, {'name': '原创'}, {'name': 'JK'}, {'name': '事後'}]})

tmpDB.search(Q['tags'].any((Q['name'].matches('J'))))  # search  tag:J
tmpDB.search(
    (Q['tags'].any((Q['name'].matches('J')))) & (Q['tags'].any((Q['name'].matches('K')))))  # search tag:J & tag:K
tmpDB.search((Q['tags'].any((Q['name'].matches('J')))) & (Q['tags'].any((Q['name'].matches('K')))) & (
    Q['tags'].any((Q['name'].matches('P')))))
# search  tag:J & tag:K & tag:P

```


I hope to support this feature, it seems that I can only write like this now....

**Comment by msiemens** [comment]
I've implemented this in bf039e8ee78fb80ba56530bedfba931eba5f2d60. It adds a `fragment` query operation:

```python
>>> db.search(Query().fragment({'foo': True, 'bar': False}))
[{'foo': True, 'bar': False, 'foobar: 'yes!'}]
>>> db.search(Query().field.fragment({'foo': True, 'bar': False}))
[{'field': {'foo': True, 'bar': False, 'foobar: 'yes!'}]
```

Would this work? 🙃

**Comment by DylanHolmesDH** [comment]
@msiemens When utilizing this feature in the latest version, I copy your example above in a sample database with just a few entries. And I get the following error:
https://pastebin.com/yn5hVeV2

It is saying an Empty query was evaluated, I am stuck and not quite sure what / if I am doing something wrong.

**Comment by msiemens** [comment]
@DylanHolmesDH This is not released right now. If you want to test this feature, you currently have to use the latest development version from GitHub (`pip install git+https://github.com/msiemens/tinydb@master`). Does this work for you?

That being said, I'll release a new version that includes this feature in the next couple of days 🙂

**Comment by msiemens** [comment]
This feature has now been released in TinyDB v4.4 🙃

**Comment by msiemens** [comment]
I'm closing this issue as the feature has been released. If you have any issues with `fragment` queries, feel free to comment or open a new issue
